### PR TITLE
feat: add a function test for read/write throttling

### PR DESCRIPTION
**Pull request by Smityz**
Add a function test for throttle

### CODE REVIEW

File: src/test/function_test/utils.h [hunk 1]
```diff
@@ -30,3 +33,50 @@
...
+                           const std::vector<std::string> values)
+{
+    std::map<std::string, std::string> result;
+    dassert(sortkeys.size() == values.size(), "sortkeys.size() != values.size()");
+    int len = sortkeys.size();
+    for (int i = 0; i < len; i++) {
+        result.emplace(std::make_pair(sortkeys[i], values[i]));
+    }
+    return result;
+}
```
Comment (levy5307) [inline src/test/function_test/utils.h:82]: add a blank line

File: src/test/function_test/utils.h [hunk 2]
```diff
@@ -30,3 +33,50 @@
...
+
+inline std::map<std::string, std::string>
+generate_sortkey_value_map(const std::vector<std::string> sortkeys,
+                           const std::vector<std::string> values)
+{
+    std::map<std::string, std::string> result;
+    dassert(sortkeys.size() == values.size(), "sortkeys.size() != values.size()");
```
Comment (levy5307) [inline src/test/function_test/utils.h:76]: dcheck_eq

File: src/test/function_test/test_throttle.cpp [hunk 3]
```diff
@@ -0,0 +1,790 @@
...
+    {
+        test_name = test_case;
+        start_time_ms = dsn_now_ms();
+        duration_ms = time_duration_s * 1000;
+        records.emplace(std::make_pair("duration_ms", duration_ms));
+    }
+
+    bool is_time_up()
+    {
+        if (dsn_now_ms() - start_time_ms > duration_ms) {
```
Comment (levy5307) [inline src/test/function_test/test_throttle.cpp:106]: return dsn_now_ms() - start_time_ms > duration_ms

File: src/test/function_test/test_throttle.cpp [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+            return true;
+        }
+        return false;
+    }
+
+    void record(uint64_t size, bool is_reject)
+    {
+        if (is_time_up()) {
+            return;
+        }
```
Comment (levy5307) [inline src/test/function_test/test_throttle.cpp:116]: auto now_ns = dsn_now_ms();
if (now_ns - start_time_ms <=10)
...

File: src/test/function_test/utils.h (hunk 2 above)
Comment (Smityz) [reply src/test/function_test/utils.h:76]: It's util, not a test

File: src/test/function_test/test_throttle.cpp [hunk 5]
```diff
@@ -0,0 +1,790 @@
...
+        if (dsn_now_ms() - start_time_ms <= 10) {
```
Comment (levy5307) [inline src/test/function_test/test_throttle.cpp:117]:
```
switch(now_ms - start_time_ms) {
case 0...10:
case 11...100:
```

File: src/test/function_test/test_throttle.cpp [hunk 6]
```diff
@@ -0,0 +1,790 @@
...
+    set,
+    multi_set
+};
+
+struct throttle_test_plan
+{
+    std::string test_plan_case;
+    operation_type ot;
+    int single_value_sz;
+    int multi_count;
```
Comment (levy5307) [inline src/test/function_test/test_throttle.cpp:58]: better to give them init value

File: src/test/function_test/test_throttle.cpp [hunk 7]
```diff
@@ -0,0 +1,790 @@
...
+        chdir(global_env::instance()._pegasus_root.c_str());
+        system("./run.sh clear_onebox");
+        system("./run.sh start_onebox -w");
+        chdir(global_env::instance()._working_dir.c_str());
+    }
+
+    void set_throttle(throttle_type type, uint64_t value)
+    {
+        std::vector<std::string> keys, values;
+        if (type == throttle_type::read_by_qps) {
```
Comment (levy5307) [inline src/test/function_test/test_throttle.cpp:199]: I think switch is better here

File: src/test/function_test/utils.h (hunk 2 above)
Comment (levy5307) [reply src/test/function_test/utils.h:76]: you can take a look at dcheck_eq in fmt_logging.h

File: src/test/function_test/test_throttle.cpp (hunk 5 above)
Comment (Smityz) [reply src/test/function_test/test_throttle.cpp:117]: I think it's just code style differences, no need to change

Comment (levy5307) [reply src/test/function_test/test_throttle.cpp:117]: I don't agree with you. Writing too many `if...else` is bad smell

File: src/test/function_test/utils.h [hunk 8]
```diff
@@ -30,3 +33,50 @@
...
+                                   "1234567890");
+    std::string result;
+    for (int i = 0; i < str_len; i++) {
+        result += chars[dsn::rand::next_u32(chars.size())];
+    }
+    return result;
+}
+
+inline std::string
+generate_hash_key_with_hotkey(bool is_hotkey, int probability = 100, uint32_t str_len = 20)
```
Comment (levy5307) [inline src/test/function_test/utils.h:50]: What about generate_hotkey?

File: src/test/function_test/utils.h [hunk 9]
```diff
@@ -30,3 +33,50 @@
             }                                                                                      \
         }                                                                                          \
     } while (0)
+
+inline std::string generate_random_str(uint32_t str_len = 20)
```
Comment (levy5307) [inline src/test/function_test/utils.h:37]: generate_random_string

File: src/test/function_test/test_throttle.cpp [hunk 10]
```diff
@@ -0,0 +1,785 @@
...
+                                                 ? dsn::rand::next_u32(test_plan.single_value_sz)
+                                                 : test_plan.single_value_sz);
+            auto sortkey_value_pairs = generate_sortkey_value_map(
+                generate_str_vector_by_random(test_sortkey_len, test_plan.multi_count),
+                generate_str_vector_by_random(
+                    test_plan.single_value_sz, test_plan.multi_count, test_plan.random_value_size));
+
+            if (test_plan.ot == operation_type::set) {
+                ref_count++;
+                pg_client->async_set(
```
Comment (levy5307) [inline src/test/function_test/test_throttle.cpp:249]: I think you can encapsulate different operation into several small functions

File: src/test/function_test/test_throttle.cpp (hunk 5 above)
Comment (Smityz) [reply src/test/function_test/test_throttle.cpp]: switch in C++ can't support `>` or `<`

File: src/test/function_test/utils.h [hunk 11]
```diff
@@ -30,3 +33,50 @@
...
+{
+    static const std::string chars("abcdefghijklmnopqrstuvwxyz"
+                                   "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                                   "1234567890");
+    std::string result;
+    for (int i = 0; i < str_len; i++) {
+        result += chars[dsn::rand::next_u32(chars.size())];
+    }
+    return result;
+}
```
Comment (Shuo-Jia) [inline src/test/function_test/utils.h:47]: consider add the function into `dsn::rand` as `common` function

File: src/test/function_test/test_throttle.cpp (hunk 10 above)
Comment (Smityz) [reply src/test/function_test/test_throttle.cpp:249]: It's difficult to capture all the related parameters like `throttle_test_recorder` `is_running` `ref_count`

File: src/test/function_test/utils.h (hunk 11 above)
Comment (Smityz) [reply src/test/function_test/utils.h:47]: I will do it in the next PR

File: src/test/function_test/utils.h (hunk 2 above)
Comment (levy5307) [reply src/test/function_test/utils.h:75]: Why did you resolve this conversation without any modification?

Comment (Smityz) [reply src/test/function_test/utils.h:75]: https://github.com/apache/incubator-pegasus/pull/842#discussion_r757346702
I will refactor these codes in the next PR, sorry for not telling

Comment (Smityz) [reply src/test/function_test/utils.h]: Have fixed

**Review by Shuo-Jia — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]

**Pull request merged**